### DISCUSSION TITLE
NetCDF output writer error

### DISCUSSION THREAD

**Discussion by Lichriszz** [CliMA/Oceananigans.jl#1408] [Computational science!]
Hi,

When I try to run this example, I got an error message (see below). Can anyone help?
Thanks!

```julia
using Oceananigans, Oceananigans.OutputWriters

grid = RegularCartesianGrid(size=(16, 16, 16), extent=(1, 1, 1));

model = IncompressibleModel(grid=grid);

simulation = Simulation(model, Δt=12, stop_time=3600);

fields = Dict("u" => model.velocities.u, "T" => model.tracers.T);

simulation.output_writers[:field_writer] =
    NetCDFOutputWriter(model, fields, filename="fields.nc", time_interval=60)

```

```
**Error message:**

LoadError: UndefKeywordError: keyword argument filepath not assigned
in expression starting at /Users/lichriszz/Desktop/Research/netcdf test.jl:11
(::Core.var"#Type##kw")(::NamedTuple{(:filename, :time_interval),Tuple{String,Int64}}, ::Type{NetCDFOutputWriter}, ::IncompressibleModel{Oceananigans.TimeSteppers.QuasiAdamsBashforth2TimeStepper{Float64,NamedTuple{(:u, :v, :w, :T, :S),Tuple{Field{Face,Center,Center,OffsetArrays.OffsetArray{Float64,3,Array{Float64,3}},RegularCartesianGrid{Float64,Periodic,Periodic,Bounded,OffsetArrays.OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}},NamedTuple{(:x, :y, 
```

**Comment by Lichriszz** [comment]
Also, I don't think the initial condition is specified in this example. I am not sure how this simulation works.

**Reply by navidcy** [reply]
Hi @Lichriszz,

1) could you clarify where you got this example from?

2) can you printout your `versioninfo()` output?

**Reply by Lichriszz** [reply]
Hi,

Thanks for the reply.
I got the example from https://clima.github.io/OceananigansDocumentation/v0.39.0/library/#Output-writers.
And here is the version information.

Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, icelake-client)
Environment:
  JULIA_NUM_THREADS = 4
  JULIA_PKG_SERVER = pkg.juliahub.com
  JULIA_DEPOT_PATH = /Users/lichriszz/.julia:/Applications/JuliaPro-1.5.3-1.app/Contents/Resources/julia/Contents/Resources/julia/local/share/julia:/Applications/JuliaPro-1.5.3-1.app/Contents/Resources/julia/Contents/Resources/julia/share/julia
  JULIA_LOAD_PATH = /Users/lichriszz/.julia/environments/JuliaPro_v1.5.3-1:@:@v#.#:@stdlib

**Reply by navidcy** [reply]
Also output of `using Pkg; Pkg.status()`? :)

**Reply by Lichriszz** [reply]
Status `~/.julia/environments/JuliaPro_v1.5.3-1/Project.toml`
  [c52e3926] Atom v0.12.25 ⚲
  [7073ff75] IJulia v1.23.1
  [033835bb] JLD2 v0.3.3
  [e5e0dc1b] Juno v0.8.4 ⚲
  [9e8cae18] Oceananigans v0.47.1
  [4722fa14] PkgAuthentication v1.1.0
  [91a5bcdd] Plots v1.10.3
  [44d3d7a6] Weave v0.10.6

**Comment by navidcy** [comment]
Seems like you are using Oceananigans v0.47.1 but looking at the documentation of version v0.39.0. Is this correct? Something might have changed in between the two versions...

Have a look at here: https://clima.github.io/OceananigansDocumentation/v0.47.1/library/#Output-writers 
I feel the examples have been updated.

Let me know if you still have issues.

**Reply by Lichriszz** [reply]
But this particular examples in the v0.47.1 library is the same.

**Reply by navidcy** [reply]
Fair point... I just manage to run the example using v0.47.1!
I don't know what might be happening. Do you run this from the REPL?

**Reply by Lichriszz** [reply]
No, I ran it normally from the editing space. But thank you so much!

**Reply by navidcy** [reply]
What's the "editing space"?

**Reply by navidcy** [reply]
<img width="780" alt="Screen Shot 2021-03-01 at 5 37 34 pm" src="https://user-images.githubusercontent.com/7112768/109460555-d8818180-7ab4-11eb-8571-589dbcb73fe4.png">

**Reply by navidcy** [reply]
perhaps @ali-ramadhan or @glwagner have some more insight...

**Reply by ali-ramadhan** [reply]
@Lichriszz Sorry for the confusion.

Like @navidcy pointed out, I think we changed the name of the keyword argument from `filename` to `filepath` for `NetCDFOutputWriter` at some point recently but v0.47.1 should be using `filepath`.

Are you sure the v0.47.1 docs use `filename` for `NetCDFOutputWriter`? The examples at https://clima.github.io/OceananigansDocumentation/v0.47.1/library/#Oceananigans.OutputWriters.NetCDFOutputWriter-Tuple{Any,Any} seem to all use `filepath` (at least from what I can see and search).

**Reply by Lichriszz** [reply]
Oh sorry, I was looking at the wrong version. Thanks for explaining!

**Reply by ali-ramadhan** [reply]
I think Google is especially bad at linking you to the latest/current docs for Julia packages.

So many times it sent me to some ancient version of the docs so now I've started checking the version every time I open the docs from any search engine haha.

**Comment by tomchor** [comment]
I could reproduce the error here using version `0.51.0`. The problem is that the keyword options for  `NetCDFOutputWriter`  have changed. You have to change your last line to:

```julia
julia> simulation.output_writers[:field_writer] =
           NetCDFOutputWriter(model, fields, filepath="fields.nc", schedule=TimeInterval(60))
```

Basically the gist of the example is the same between `0.39` and `0.47`, but some of the keywords used changed a bit, which is why you're getting an error (and probably because you didn't notice the change, since it was very minor).

Most examples in the Oceananigans docs are routinely tested to make sure they work! So it's good to always make sure you get the examples from the version of the docs that match your installed version :)

**Reply by tomchor** [reply]
Just a note, if you want to run your example with a more recent version of Oceananigans (which I guess is advised, no?) you will also have to change the second line to

```julia
grid = RegularRectilinearGrid(size=(16, 16, 16), extent=(1, 1, 1));
```
to match the more recent docs, otherwise you'll run into a similar error.
